fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java

```diff
@@ -0,0 +1,217 @@
+import com.apple.foundationdb.record.provider.foundationdb.properties.RecordLayerPropertyStorage;
+import com.apple.foundationdb.tuple.Tuple;
+import com.apple.test.Tags;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import java.util.concurrent.ExecutionException;
+import java.util.stream.IntStream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * Tests for AgilityContext.
+ */
+@Tag(Tags.RequiresFDB)
+class AgilityContextTest extends FDBRecordStoreTestBase {
+    int loopCount = 20;
+    int threadCount = 5; // if exceeds a certain size, may cause an execution pool deadlock
+
+    void testAgilityContextConcurrentSingleObject(final AgilityContext agilityContext, boolean doFlush) throws ExecutionException, InterruptedException {
+        agilityContextTestSingleThread(0, agilityContext, doFlush);
+        agilityContextTestSingleThread(1, agilityContext, doFlush);
+        agilityContextTestSingleThread(1, agilityContext, doFlush);
+
+        for (int loop = 0; loop < loopCount; loop++) {
+            final int loopIndex = loop;
+            IntStream.rangeClosed(0, threadCount).parallel().forEach(i -> {
+                try {
+                    agilityContextTestSingleThread(loopIndex * i, agilityContext, doFlush);
+                } catch (ExecutionException | InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+        }
+    }
+
+    private void agilityContextTestSingleThread(int i, AgilityContext agilityContext, boolean explicityFlush) throws ExecutionException, InterruptedException {
+        byte[] key = Tuple.from(500, i).pack();
+        byte[] val = Tuple.from(i).pack();
+        try {
+            agilityContext.set(key, val);
+            if (explicityFlush && i % 3 == 0) { // keep some un-flushed values
+                final byte[] bytes = agilityContext.get(key).get();
+                long readVal = Tuple.fromBytes(bytes).getLong(0); // maybe flushed, probably not
+                assertEquals(readVal, i);
+                agilityContext.flush();
+                readVal = Tuple.fromBytes(bytes).getLong(0); // surely flushed
+                assertEquals(readVal, i);
+            }
+        } catch (ExecutionException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void assertLoopThreadsValues() {
+        try (FDBRecordContext context = fdb.openContext()) {
+            for (int loop = 0; loop < loopCount; loop++) {
+                final AgilityContext agilityContext = AgilityContext.factory(context, false);
+                for (int thread = 0; thread < threadCount; thread++) {
+                    int i = loop * thread;
+                    byte[] key = Tuple.from(500, i).pack();
+                    byte[] val = Tuple.from(i).pack();
```

Comment:
These overwrite each other (e.g. loop=2 and thread=2 write the same value to the same key).
Is that intentional, or would it be better to have `Tuple.from(500, loop, thread)` and `Tuple.from(loop, thread)` ?

Reply:
I've tried to create some arbitrary overlap to ensure no conflicts. But I can change it to your suggestion.

Reply:
✅